roachtest: generalize version-upgrade to per-node op

tbg opened the pull request:
First three commits from #47268, which is borsing right now.

----

In anticipation of tests that want to use mixed version testing on
clusters that contain workload-only nodes, set up the harness so that
this is natively supported.

Release note: None

tbg committed:
roachtest: refactor feature tests in version-upgrade

It had become clear that they were built in too deeply into the mixed
version testing harness and could simply be replaced by something using
the existing concept of a step.
We now pepper the step list with the feature testing step manually,
which is more in line with how I expect this framework to be used
in general, and also avoids running them between steps where nothing
really changed.

Release note: None

tbg committed:
roachtest: inline single-caller step in version-upgrade

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47275)
<!-- Reviewable:end -->

tbg committed:
roachtest: generalize version-upgrade to per-node ops

In anticipation of tests that want to use mixed version testing on
clusters that contain workload-only nodes, set up the harness so that
this is natively supported.

Release note: None

spaskob reviewed (approved):
Thanks Tobi, this looks great!

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47275#-:-M4UZNfV6g-lmUzaqLLu:bx5oil6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spaskob)



<!-- Sent from Reviewable.io -->

tbg commented:
cc @otan just for awareness

tbg committed:
roachtest: skip version-upgrade again

I thought the [bug] would not repro after I bumped the base fixture, but
it still does (though maybe one in fifteen only) and I'm running out of
work day, so skip the test again. Tomorrow's a latter day.

[bug]: https://github.com/cockroachdb/cockroach/issues/44732#issuecomment-611484831

Release note: None

tbg commented:
Had to skip the test again, as I managed to repro the problem it was originally skipped for. That's good, in a sense, as it will give me a way to confirm that this is really the bug I still have to fix.

bors r=spaskob
